lib/ex_unit/lib/ex_unit/assertions.ex

```diff
@@ -357,6 +362,14 @@ defmodule ExUnit.Assertions do
   end
 
   @doc false
+  def __match__({:when, _, _} = left, _, _, _, _) do
+    raise ArgumentError, """
+    invalid pattern in assert/1: #{Macro.to_string(left)}
+
+    Use assert match?(... when ...) if you need to assert with a guard.
```

Comment:
We could even print out the actual code they were trying to use, right? We have all the ASTs and stuff. Maybe

```elixir
Use this if you need to assert with a guard:

    assert match?(actual_code)
```

Reply:
I like it, but I can see how it could end up with a big suggestion if the pattern is a huge map for instance (and `Macro.to_string/2` doesn't have a limit like inspect).
What do you think?

Reply:
I’m not sure someone would assert on a huge map with a guard at the end, as in I’m not sure that's a common pattern. If that ends up happening, you could always format the code to make it fit nicely.

Reply:
Makes sense.
In that case, perhaps I can remove the `#{Macro.to_string(left)}` above which would otherwise be redundant?